### PR TITLE
[GEN] Extend Matrix2DBlockLoadOp OCL usages

### PR DESCRIPTION
**Pull request by whitneywhtsang**

### CODE REVIEW

File: test/Conversion/intel/tritongpu_to_llvm_intel_block_ptr.mlir [hunk 1]
```diff
@@ -3,8 +3,8 @@
 module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 32 : i32, triton_gpu.shared = 0 : i32, "triton_gpu.threads-per-warp" = 1 : i32} {
   // CHECK-DAG: llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockWrite.v8i32(i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32, vector<8xi32>)
   // CHECK-DAG: llvm.func spir_funccc @_Z38intel_sub_group_f16_f16_matrix_mad_k16Dv8_sDv8_iDv8_f(vector<8xi16>, vector<8xi32>, vector<8xf32>) -> vector<8xf32> attributes {passthrough = ["convergent"]}
-  // CHECK-DAG: llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockRead.v32i32(i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32) -> vector<32xi32>
-  // CHECK-DAG: llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockRead.v64i16(i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32) -> vector<64xi16>
+  // CHECK-DAG: llvm.func spir_funccc @_Z42intel_sub_group_2d_block_read_16b_32r16x2cPU3AS1viiiDv2_iPt(!llvm.ptr<1>, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["convergent"]}
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/Conversion/intel/tritongpu_to_llvm_intel_block_ptr.mlir:6]: Addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1278.

File: test/Conversion/intel/tritongpu_to_llvm_intel_block_ptr.mlir [hunk 2]
```diff
@@ -3,8 +3,8 @@
...
+  // CHECK-DAG: llvm.func spir_funccc @_Z52intel_sub_group_2d_block_read_transform_16b_32r16x2cPU3AS1viiiDv2_iPj(!llvm.ptr<1>, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["convergent"]}
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/Conversion/intel/tritongpu_to_llvm_intel_block_ptr.mlir:7]: Addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1278.

File: test/TritonGEN/tritongen-2Dblockload-to-llvm.mlir [hunk 3]
```diff
@@ -0,0 +1,307 @@
+// RUN: triton-opt -convert-tritongen-to-llvm -split-input-file %s | FileCheck %s
+
+// CHECK: llvm.func spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x1cPU3AS1viiiDv2_iPt(!llvm.ptr<1>, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["convergent"]}
+
+llvm.func @triton_gen.2Dblockload(%ptr : !llvm.ptr<1>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonGEN/tritongen-2Dblockload-to-llvm.mlir:5]: Limited the number of CHECKs except the first test.

**Comment by whitneywhtsang** [comment]
This change depends on coming agama rolling driver.

**Review by jopperm — APPROVED** [review]

File: test/Conversion/intel/tritongpu_to_llvm_intel_block_ptr.mlir [hunk 4]
```diff
@@ -3,8 +3,8 @@
 module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 32 : i32, triton_gpu.shared = 0 : i32, "triton_gpu.threads-per-warp" = 1 : i32} {
   // CHECK-DAG: llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockWrite.v8i32(i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32, vector<8xi32>)
   // CHECK-DAG: llvm.func spir_funccc @_Z38intel_sub_group_f16_f16_matrix_mad_k16Dv8_sDv8_iDv8_f(vector<8xi16>, vector<8xi32>, vector<8xf32>) -> vector<8xf32> attributes {passthrough = ["convergent"]}
-  // CHECK-DAG: llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockRead.v32i32(i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32) -> vector<32xi32>
-  // CHECK-DAG: llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockRead.v64i16(i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32) -> vector<64xi16>
+  // CHECK-DAG: llvm.func spir_funccc @_Z42intel_sub_group_2d_block_read_16b_32r16x2cPU3AS1viiiDv2_iPt(!llvm.ptr<1> {llvm.nonnull}, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["nounwind"]}
```
Comment (etiotto) [inline test/Conversion/intel/tritongpu_to_llvm_intel_block_ptr.mlir:6]: This needs passthrough attribute to describe the memory semantics (function can only read the memory pointed to by the ptr argument). Applies to all similar cases.

File: test/TritonGEN/tritongen-2Dblockload-to-llvm.mlir [hunk 5]
```diff
@@ -0,0 +1,271 @@
+// RUN: triton-opt -convert-tritongen-to-llvm -split-input-file %s | FileCheck %s
+
+// CHECK: llvm.func spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x1cPU3AS1viiiDv2_iPt(!llvm.ptr<1> {llvm.nonnull}, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["nounwind"]}
```
Comment (etiotto) [inline test/TritonGEN/tritongen-2Dblockload-to-llvm.mlir:3]: Add passthrough attribute for mem semantics

File: test/TritonGEN/tritongen-invalid.mlir [hunk 6]
```diff
@@ -235,6 +235,14 @@ llvm.func @matrix_2Dblockload(%ptr : !llvm.ptr, %base_width : i32, %base_height
 
 // -----
 
+llvm.func @matrix_2Dblockload(%ptr : !llvm.ptr, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
+  // expected-error @+1 {{'triton_gen.2Dblockload' op expecting tile_width to be 16 with vnni transform}}
```
Comment (etiotto) [inline test/TritonGEN/tritongen-invalid.mlir:239]: .. equal to 16 when vnni_transform is true

File: test/TritonGEN/tritongen-to-llvm.mlir [hunk 7]
```diff
@@ -297,48 +297,6 @@ llvm.func @triton_gen.dpas.f32(%c : vector<8xf32>, %a : vector<4xf32>, %b : vect
 
 // -----
 
-// CHECK: llvm.func spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(!llvm.ptr<1> {llvm.nonnull}, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["nounwind"]}
```
Comment (etiotto) [inline test/TritonGEN/tritongen-to-llvm.mlir:300]: What happened to this tests, where they moved?

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 8]
```diff
@@ -48,26 +48,32 @@ template <typename Op> static LogicalResult verifyInput(Op op) {
...
+                               "to systolic depth times 2, i.e., 16 elements");
+      break;
+    case 8:
+      if (TileWidth != 32)
+        return op->emitOpError("tile_width for 8 bit elements should be equal "
+                               "to systolic depth times 4, i.e., 32 elements");
+      break;
+    default:
+      return op->emitOpError("element size should be 8, 16 or 32 bits");
+    }
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:76]:
``` 
if (op.getVnniTransform() && TileWidth != 16)) {
  return ... 
  
if (!op.getVnniTransform())
  switch....

File: test/TritonGEN/tritongen-to-llvm.mlir (hunk 7 above)
Comment (whitneywhtsang) [reply test/TritonGEN/tritongen-to-llvm.mlir:300]: moved to `tritongen-2Dblockload-to-llvm.mlir` which contains all variants.

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp (hunk 8 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp]: Done in https://github.com/intel/intel-xpu-backend-for-triton/pull/1290.

File: test/TritonGEN/tritongen-invalid.mlir (hunk 6 above)
Comment (whitneywhtsang) [reply test/TritonGEN/tritongen-invalid.mlir]: Done in https://github.com/intel/intel-xpu-backend-for-triton/pull/1290.

File: test/TritonGEN/tritongen-2Dblockload-to-llvm.mlir (hunk 5 above)
Comment (whitneywhtsang) [reply test/TritonGEN/tritongen-2Dblockload-to-llvm.mlir]: Done in https://github.com/intel/intel-xpu-backend-for-triton/pull/1224.

File: test/Conversion/intel/tritongpu_to_llvm_intel_block_ptr.mlir (hunk 4 above)
Comment (whitneywhtsang) [reply test/Conversion/intel/tritongpu_to_llvm_intel_block_ptr.mlir]: Done in https://github.com/intel/intel-xpu-backend-for-triton/pull/1224.

**Review by etiotto — APPROVED** [review]

**Pull request merged**